packages/flutter_tools/lib/src/version.dart

```diff
@@ -24,7 +24,7 @@ enum Channel {
 }
 
 /// The flutter GitHub repository.
-const String _flutterGit = 'https://github.com/flutter/flutter.git';
+final String _flutterGit = globals.platform.environment['FLUTTER_GIT_URL'] ?? 'https://github.com/flutter/flutter.git';
```

Comment:
This looks like a good solution. One thing though, is that if the user has a typo in this env variable (or if it the remote is down), we should explicitly show the override value in any error message. If you check line 654 of this file, we throw an error if a command we run does not return a 0 exit code. Even though this error includes the complete command that failed, unfortunately, if `FLUTTER_GIT_URL` env variable is wrong, we won't hit an error until line 257, which tries to run the command `git fetch $_versionCheckRemote $branch`, which doesn't include the upstream URL in it. Probably the better version would be to add a `catch` around 262, and IF we have an explicit `FLUTTER_GIT_URL` override set in the environment, `globals.printError('Warning: the Flutter git upstream was overriden by the environment variable FLUTTER_GIT_URL = ${globals.platform.environment['FLUTTER_GIT_URL']}');` (something like that, I didn't test it).

Also, this change should have unit tests in the file: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/version_test.dart. Using the `testUsingContext()` function, you can provide a `FakePlatform` and `ProcessManager` overrides (if you search the codebase for `FakePlatform`, you'll find a lot of examples) where you set the environment variable and then expect that you are getting the correct git commands to the `mockProcessManager`, specifically the one setting the upstream.

Let me know if you you run into any obstacles!

Reply:
I have finished print warning and unit test. But test running failed unexpected. Sorry for troubling but I have a problem now.